src/main/java/com/powsybl/openloadflow/sa/AbstractSecurityAnalysis.java

```diff
@@ -0,0 +1,223 @@
+        Objects.requireNonNull(workingVariantId);
+        Objects.requireNonNull(securityAnalysisParameters);
+        Objects.requireNonNull(contingenciesProvider);
+        return CompletableFuture.supplyAsync(() -> {
+            String oldWorkingVariantId = network.getVariantManager().getWorkingVariantId();
+            network.getVariantManager().setWorkingVariant(workingVariantId);
+            SecurityAnalysisResult result = runSync(securityAnalysisParameters, contingenciesProvider);
+            network.getVariantManager().setWorkingVariant(oldWorkingVariantId);
+            return result;
+        });
+    }
+
+    abstract SecurityAnalysisResult runSync(SecurityAnalysisParameters securityAnalysisParameters, ContingenciesProvider contingenciesProvider);
+
+    /**
+     * Detect violations on branches and on buses
+     * @param branches branches on which the violation limits are checked
+     * @param buses buses on which the violation limits are checked
+     * @param violations list on which the violation limits encountered are added
+     */
+    protected void detectViolations(Stream<LfBranch> branches, Stream<LfBus> buses, Map<Pair<String, Branch.Side>, LimitViolation> violations) {
+        // Detect violation limits on branches
+        branches.forEach(branch -> detectBranchViolations(branch, violations));
+
+        // Detect violation limits on buses
+        buses.forEach(bus -> detectBusViolations(bus, violations));
+    }
+
+    /**
+     * Detect violation limits on one branch and add them to the given list
+     * @param branch branch of interest
+     * @param violations list on which the violation limits encountered are added
+     */
+    protected void detectBranchViolations(LfBranch branch, Map<Pair<String, Branch.Side>, LimitViolation> violations) {
+        // detect violation limits on a branch
+        if (branch.getBus1() != null) {
+            branch.getLimits1(LimitType.CURRENT).stream()
+                .filter(temporaryLimit1 -> branch.getI1().eval() > temporaryLimit1.getValue())
+                .findFirst()
+                .map(temporaryLimit1 -> createLimitViolation1(branch, temporaryLimit1, LimitViolationType.CURRENT, PerUnit.SB / branch.getBus1().getNominalV(), branch.getI1().eval()))
+                .ifPresent(limitViolation -> violations.put(getSubjectSideId(limitViolation), limitViolation));
+
+            branch.getLimits1(LimitType.ACTIVE_POWER).stream()
+                  .filter(temporaryLimit1 -> branch.getP1().eval() > temporaryLimit1.getValue())
+                  .findFirst()
+                  .map(temporaryLimit1 -> createLimitViolation1(branch, temporaryLimit1, LimitViolationType.ACTIVE_POWER, PerUnit.SB, branch.getP1().eval()))
+                  .ifPresent(limitViolation -> violations.put(getSubjectSideId(limitViolation), limitViolation));
+
+            double apparentPower1 = computeApparentPower(branch.getP1().eval(), branch.getQ1().eval());
```

Comment:
This could not be a method inside the branch?

Reply:
You see this method as a part of LfBranch interface ?

Reply:
In general, I put some methods in the `LfBranch` interface to deal with special branches such as switch, dangling lines and legs. If you leave this method here, because it is more efficient, don't forget to think about how the special branches behave.